bin/cli.rs

```diff
@@ -0,0 +1,126 @@
+                    .unwrap_or_else(|_| panic!("Invalid RPC URL supplied: {rpc_url}")),
+            );
+
+            // Retrieve block to snapshot.
+            let runtime = runtime::Runtime::new().unwrap();
+            let block = runtime
+                .block_on(provider.get_block(block_id, BlockTransactionsKind::Full))
+                .unwrap_or_else(|e| panic!("Failed to fetch block from provider: {e}",))
+                .unwrap_or_else(|| panic!("No block found for ID: {:?}", block_id));
+
+            // TODO: We could support more specs here. This could be another CLI argument.
+            // TODO: There is a a weak link between spec_id, provider and thus RPCStorage.
+            // Not all EVM chains support all hardforks, especially rollups.
+            let chain = PevmEthereum::mainnet();
+            let spec_id = chain.get_block_spec(&block.header).unwrap();
+            let storage =
+                RpcStorage::new(provider, spec_id, BlockId::number(block.header.number - 1));
+
+            // Populate RPC storage.
+            let concurrency_level = thread::available_parallelism().unwrap_or(NonZeroUsize::MIN);
+            let mut pevm = Pevm::default();
+            let _ = pevm
+                .execute(&storage, &chain, block.clone(), concurrency_level, false)
+                .unwrap();
+
+            // Create dir.
+            let dir = format!("{output_dir}/blocks/{}", block.header.number);
+            create_dir_all(dir.clone())
+                .unwrap_or_else(|e| panic!("Failed to create directory: {e}"));
+
+            // Create blockfile.
+            let block_file = File::create(format!("{dir}/block.json"))
+                .unwrap_or_else(|e| panic!("Failed to create block file: {e}"));
+            serde_json::to_writer(block_file, &block).expect("Failed to write block to file");
+
+            let mut state = BTreeMap::<Address, EvmAccount>::new();
+            let mut bytecodes: BTreeMap<B256, EvmCode> = if let Some(path) = bytecode_path {
+                let file = File::open(path)
+                    .unwrap_or_else(|e| panic!("Failed to open bytecode file {e}",));
+                bincode::deserialize_from(BufReader::new(file))
+                    .unwrap_or_else(|e| panic!("Failed to decode bytecode file {e}"))
+            } else {
+                BTreeMap::new()
+            };
+
+            bytecodes.extend(storage.get_cache_bytecodes());
+            for (address, mut account) in storage.get_cache_accounts() {
+                if let Some(code) = account.code.take() {
+                    assert_ne!(account.code_hash.unwrap(), KECCAK_EMPTY);
+                    bytecodes.insert(account.code_hash.unwrap(), code);
+                }
+                state.insert(address, account);
+            }
+
+            let file_state = File::create(format!("{dir}/pre_state.json")).unwrap();
+            // Convert BtreeMap to serde IndexMap to provide consistent ordering.
```

Comment:
I've added this to attempt to enforce the ordering for ease of diffs. It as behaving deterministically on my machine, but may need someone else to pull this down to verify.

Reply:
`preserve_order` only works if we snapshot with sequential execution. Parallel state access can be non-deterministic for complicated blocks. 

Let's keep this as we should do sequential execution for stability anyway. The bottleneck is networking so trading stability for parallel performance isn't worth it.